In [8]:
import pandas as pd
import os

os.getcwd()

'/Users/janhavisingh/personal_projects/retail-personalisation-system/notebooks'

### 1. Load Data

In [10]:
customers = pd.read_csv("/Users/janhavisingh/personal_projects/retail-personalisation-system/data/raw/customers.csv")
articles = pd.read_csv("/Users/janhavisingh/personal_projects/retail-personalisation-system/data/raw/articles.csv")
sample_submission = pd.read_csv("/Users/janhavisingh/personal_projects/retail-personalisation-system/data/raw/sample_submission.csv")
transactions_train = pd.read_csv("/Users/janhavisingh/personal_projects/retail-personalisation-system/data/raw/transactions_train.csv")


### 2. Dataset Overview

#### Customers dataset

In [28]:
customers.shape

(1371980, 7)

In [29]:
customers.head()

,customer_id,FN,Active,club_member_status,fashion_news_frequency,age,postal_code
0,00000dbacae5abe5e23885899a1fa44253a17956c6d1c3...,NaN,NaN,ACTIVE,NONE,49.0,52043ee2162cf5aa7ee79974281641c6f11a68d276429a...
1,0000423b00ade91418cceaf3b26c6af3dd342b51fd051e...,NaN,NaN,ACTIVE,NONE,25.0,2973abc54daa8a5f8ccfe9362140c63247c5eee03f1d93...
2,000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...,NaN,NaN,ACTIVE,NONE,24.0,64f17e6a330a85798e4998f62d0930d14db8db1c054af6...
3,00005ca1c9ed5f5146b52ac8639a40ca9d57aeff4d1bd2...,NaN,NaN,ACTIVE,NONE,54.0,5d36574f52495e81f019b680c843c443bd343d5ca5b1c2...
4,00006413d8573cd20ed7128e53b7b13819fe5cfc2d801f...,1.0,1.0,ACTIVE,Regularly,52.0,25fa5ddee9aac01b35208d01736e57942317d756b32ddd...


In [30]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1371980 entries, 0 to 1371979
Data columns (total 7 columns):
 #   Column                  Non-Null Count    Dtype  
---  ------                  --------------    -----  
 0   customer_id             1371980 non-null  object 
 1   FN                      476930 non-null   float64
 2   Active                  464404 non-null   float64
 3   club_member_status      1365918 non-null  object 
 4   fashion_news_frequency  1355969 non-null  object 
 5   age                     1356119 non-null  float64
 6   postal_code             1371980 non-null  object 
dtypes: float64(3), object(4)
memory usage: 73.3+ MB


In [31]:
customers.duplicated().sum()

np.int64(0)

In [64]:
customers_missing = pd.DataFrame({
    "Missing Count": customers.isna().sum(),
    "Missing Percentage (%)": (customers.isna().mean() * 100).round(2)
})
customers_missing

,Missing Count,Missing Percentage (%)
customer_id,0,0.00
FN,895050,65.24
Active,907576,66.15
club_member_status,6062,0.44
fashion_news_frequency,16011,1.17
age,15861,1.16
postal_code,0,0.00


In [45]:
print("FN")
print(customers["FN"].describe())
print()
print("Active")
print(customers["Active"].describe())
print()
print("age")
print(customers["age"].astype("Int64").describe())


FN
count    476930.0
mean          1.0
std           0.0
min           1.0
25%           1.0
50%           1.0
75%           1.0
max           1.0
Name: FN, dtype: float64

Active
count    464404.0
mean          1.0
std           0.0
min           1.0
25%           1.0
50%           1.0
75%           1.0
max           1.0
Name: Active, dtype: float64

age
count    1356119.0
mean     36.386965
std      14.313628
min           16.0
25%           24.0
50%           32.0
75%           49.0
max           99.0
Name: age, dtype: Float64


In [44]:
print(customers["club_member_status"].value_counts())
print()
print(customers["fashion_news_frequency"].value_counts())

club_member_status
ACTIVE        1272491
PRE-CREATE      92960
LEFT CLUB         467
Name: count, dtype: int64

fashion_news_frequency
NONE         877711
Regularly    477416
Monthly         842
Name: count, dtype: int64


#### Articles Dataset

In [35]:
articles.shape

(105542, 25)

In [36]:
articles.head()

,article_id,product_code,prod_name,product_type_no,product_type_name,product_group_name,graphical_appearance_no,graphical_appearance_name,colour_group_code,colour_group_name,...,department_name,index_code,index_name,index_group_no,index_group_name,section_no,section_name,garment_group_no,garment_group_name,detail_desc
0,108775015,108775,Strap top,253,Vest top,Garment Upper body,1010016,Solid,9,Black,...,Jersey Basic,A,Ladieswear,1,Ladieswear,16,Womens Everyday Basics,1002,Jersey Basic,Jersey top with narrow shoulder straps.
1,108775044,108775,Strap top,253,Vest top,Garment Upper body,1010016,Solid,10,White,...,Jersey Basic,A,Ladieswear,1,Ladieswear,16,Womens Everyday Basics,1002,Jersey Basic,Jersey top with narrow shoulder straps.
2,108775051,108775,Strap top (1),253,Vest top,Garment Upper body,1010017,Stripe,11,Off White,...,Jersey Basic,A,Ladieswear,1,Ladieswear,16,Womens Everyday Basics,1002,Jersey Basic,Jersey top with narrow shoulder straps.
3,110065001,110065,OP T-shirt (Idro),306,Bra,Underwear,1010016,Solid,9,Black,...,Clean Lingerie,B,Lingeries/Tights,1,Ladieswear,61,Womens Lingerie,1017,"Under-, Nightwear","Microfibre T-shirt bra with underwired, moulde..."
4,110065002,110065,OP T-shirt (Idro),306,Bra,Underwear,1010016,Solid,10,White,...,Clean Lingerie,B,Lingeries/Tights,1,Ladieswear,61,Womens Lingerie,1017,"Under-, Nightwear","Microfibre T-shirt bra with underwired, moulde..."


In [37]:
articles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105542 entries, 0 to 105541
Data columns (total 25 columns):
 #   Column                        Non-Null Count   Dtype 
---  ------                        --------------   ----- 
 0   article_id                    105542 non-null  int64 
 1   product_code                  105542 non-null  int64 
 2   prod_name                     105542 non-null  object
 3   product_type_no               105542 non-null  int64 
 4   product_type_name             105542 non-null  object
 5   product_group_name            105542 non-null  object
 6   graphical_appearance_no       105542 non-null  int64 
 7   graphical_appearance_name     105542 non-null  object
 8   colour_group_code             105542 non-null  int64 
 9   colour_group_name             105542 non-null  object
 10  perceived_colour_value_id     105542 non-null  int64 
 11  perceived_colour_value_name   105542 non-null  object
 12  perceived_colour_master_id    105542 non-null  int64 
 13 

In [38]:
articles.duplicated().sum()

np.int64(0)

In [39]:
articles.isna().sum()

article_id                        0
product_code                      0
prod_name                         0
product_type_no                   0
product_type_name                 0
product_group_name                0
graphical_appearance_no           0
graphical_appearance_name         0
colour_group_code                 0
colour_group_name                 0
perceived_colour_value_id         0
perceived_colour_value_name       0
perceived_colour_master_id        0
perceived_colour_master_name      0
department_no                     0
department_name                   0
index_code                        0
index_name                        0
index_group_no                    0
index_group_name                  0
section_no                        0
section_name                      0
garment_group_no                  0
garment_group_name                0
detail_desc                     416
dtype: int64

In [69]:
pd.DataFrame({
    "ProductUnique": [articles["prod_name"].nunique()],
    "SectionUnique": [articles["section_name"].nunique()],
    "DepartmentUnique": [articles["department_name"].nunique()]
})

,ProductUnique,SectionUnique,DepartmentUnique
0,45875,56,250


In [73]:
import pandas as pd

def analyze_column_summary(df, cols):
    summary_data = []
    
    for col in cols:
        vc = df[col].value_counts(normalize=True)
        top_5 = df[col].value_counts().head(5).to_dict()
        top_5_pct = (vc.head(5).sum() * 100).round(2)
        
        summary_data.append({
            "Column": col,
            "Unique Values": df[col].nunique(),
            "Top 5 Categories (Counts)": top_5,
            "Top 5 Concentration (%)": f"{top_5_pct}%"
        })
        
    return pd.DataFrame(summary_data)

target_cols = ["prod_name", "section_name", "department_name"]
result_df = analyze_column_summary(articles, target_cols)
result_df

,Column,Unique Values,Top 5 Categories (Counts),Top 5 Concentration (%)
0,prod_name,45875,"{'Dragonfly dress': 98, 'Mike tee': 72, 'Wow p...",0.33%
1,section_name,56,"{'Womens Everyday Collection': 7295, 'Divided ...",26.26%
2,department_name,250,"{'Jersey': 4604, 'Knitwear': 3503, 'Trouser': ...",14.41%


#### Transactions Train Dataset

In [52]:
transactions_train.shape

(31788324, 5)

In [53]:
transactions_train.head()

,t_dat,customer_id,article_id,price,sales_channel_id
0,2018-09-20,000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...,663713001,0.050831,2
1,2018-09-20,000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...,541518023,0.030492,2
2,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699...,505221004,0.015237,2
3,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699...,685687003,0.016932,2
4,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699...,685687004,0.016932,2


In [55]:
transactions_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31788324 entries, 0 to 31788323
Data columns (total 5 columns):
 #   Column            Dtype  
---  ------            -----  
 0   t_dat             object 
 1   customer_id       object 
 2   article_id        int64  
 3   price             float64
 4   sales_channel_id  int64  
dtypes: float64(1), int64(2), object(2)
memory usage: 1.2+ GB


In [57]:
transactions_train.duplicated().sum()

np.int64(2974905)

In [58]:
transactions_train.isna().sum()

t_dat               0
customer_id         0
article_id          0
price               0
sales_channel_id    0
dtype: int64

In [62]:
transactions_train["price"].describe()

count    3.178832e+07
mean     2.782927e-02
std      1.918113e-02
min      1.694915e-05
25%      1.581356e-02
50%      2.540678e-02
75%      3.388136e-02
max      5.915254e-01
Name: price, dtype: float64

### 3. Interaction Behaviour 

In [82]:
customer_interactions = (
    transactions_train
    .groupby("customer_id")
    .agg(
        TotalTrans=("price", "count"),
        UniqueArticle=("article_id", "nunique")
    )
)
bins = [0, 1, 2, 5, 10, 20, 50, float("inf")]
labels = ["1", "2", "3-5", "6-10", "11-20", "21-50", "51+"]

customer_interactions["TransactionBucket"] = pd.cut(
    customer_interactions["TotalTrans"],
    bins=bins,
    labels=labels
)

customer_interactions["UniqueArticleBucket"] = pd.cut(
    customer_interactions["UniqueArticle"],
    bins=bins,
    labels=labels
)
transaction_distribution = (
    customer_interactions["TransactionBucket"]
    .value_counts(normalize=True, sort=False)
    .mul(100)
)

unique_article_distribution = (
    customer_interactions["UniqueArticleBucket"]
    .value_counts(normalize=True, sort=False)
    .mul(100)
)

print(transaction_distribution)
print(unique_article_distribution)

TransactionBucket
1         9.653955
2         9.354972
3-5      17.793906
6-10     16.086329
11-20    16.046983
21-50    18.493615
51+      12.570241
Name: proportion, dtype: float64
UniqueArticleBucket
1        11.374085
2         9.747255
3-5      18.540154
6-10     16.216919
11-20    15.921678
21-50    17.834720
51+      10.365189
Name: proportion, dtype: float64


In [86]:
repeat_interactions = (
    transactions_train
    .groupby(["customer_id", "article_id"])
    .agg(TotalTrans=("price", "count"))
    .reset_index()
)
repeat_interactions["TotalTrans"].value_counts(normalize=True).sort_index().head(15) * 100

TotalTrans
1     87.126066
2     10.759228
3      1.427887
4      0.438538
5      0.105276
6      0.067471
7      0.021438
8      0.016963
9      0.008639
10     0.007680
11     0.003468
12     0.003973
13     0.001952
14     0.001842
15     0.001516
Name: proportion, dtype: float64

So repeat purchasing of the exact same article exists, but it isn't dominant. Roughly:

100 - 87.13 = 12.87\%
of observed customer–article pairs have more than one transaction.

In [89]:
customers_with_repeats = (
    repeat_interactions
    .groupby("customer_id")["TotalTrans"]
    .max()
)
customers_with_repeats.sort_values(ascending=False)

customer_id
d00063b94dcb1342869d4994844a2742b5d62927f36843164fb3f818f630bca9    570
94665b46e194622ccdbcadc0170f13a2f8ede1ff6d057d43a19b8938c808b662    199
61da44a2758206d5701771f4315637b40c8321b511191654fb1430a6408e4dfa    188
ef38ec0f0cb29ee8bbb87efc82fd16f4b99127e3eeefe69c9b5fce627e93e270    170
5cba04ed9a3759bc02a8a9e01efccc07ce76c35c1a70dce6c5cedba32ce3ab6b    166
                                                                   ... 
9ab45f030ad6a6104036223950f1236e4ea7f43747e584c810e6fe24eece972c      1
3c78f9c8d0721e8d94d959f3b17eef0f3dd13afb26d1f5c875ceceae1530fc42      1
9ab48c5bc457e6ea1a98fdb8252d25f7e1f383222ee876ab719fa8e6aacf230c      1
9ab4963ccc751829f918127fe950565207ff3fb807859e357aa9f684f354a6c7      1
ffffd9ac14e89946416d80e791d064701994755c3ab686a1eaf3458c36f52241      1
Name: TotalTrans, Length: 1362281, dtype: int64

#### Interaction Sparsity

In [81]:
total_customers = customers["customer_id"].nunique()
total_articles = articles["article_id"].nunique()

observed_combinations = (
    transactions_train[["customer_id", "article_id"]]
    .drop_duplicates()
    .shape[0]
)

total_combinations = total_customers * total_articles

density = observed_combinations / total_combinations
sparsity = 1 - density

print("Unique customers:", total_customers)
print("Unique articles:", total_articles)
print("Observed customer-article pairs:", observed_combinations)
print("Density:", density)
print("Sparsity:", sparsity)

Unique customers: 1371980
Unique articles: 105542
Observed customer-article pairs: 27306439
Density: 0.00018857840918987812
Sparsity: 0.9998114215908102


So:

Observed: only about 0.019% of all possible customer–article pairs have a purchase.

Unobserved: about 99.981% have no recorded purchase.

That's an extremely sparse interaction matrix, which is very normal for retail recommenders.

But there's an important conceptual distinction:
99.981% unobserved does not mean customers dislike 99.981% of the products.